[receiver/gitproviderreceiver] add Pull Request metrics

4lch4 opened the pull request:
**Description:**
This is the 3rd PR for the `gitproviderreceiver` that adds functionality for the following Pull Request metrics:

- `git.repository.pull_request.open.count`
- `git.repository.pull_request.open.time`
- `git.repository.pull_request.merged.count`
- `git.repository.pull_request.merged.time`
- `git.repository.pull_request.approved.time`

**Link to tracking Issue:** #22028 

**Testing:**
Basic tests included for new functionality.

**Documentation:**
Documentation has been updated to include information regarding the new metrics.

4lch4 committed:
chore: fixed spelling error in comment

4lch4 committed:
chore: formatted metadata.yaml

4lch4 committed:
feat: added new metric definitions

This commit adds the definitions for the new metrics to the metadata.yaml file. It also includes the resulting updates for the code that's auto-generated.

4lch4 committed:
feat(helpers): added `getPullRequests` method

This commit adds the `getPullRequests` method to the `helpers.go` file.

I'm making this such a small commit mostly so I can verify my GitKraken client can create a commit and sign it properly.

4lch4 committed:
feat: added 2 remaining PR metrics

When adding the code for the 4 metrics I set out to do, I realized adding the other 2 (PR Approval Time & PR Deployment Time) were really simple and easy to add so I included them as well. I can just as easily remove it if the team decides otherwise once I talk to them.

4lch4 committed:
feat: added logic for PR metrics

This commit adds the actual logic to gather & emit the PR metrics.

4lch4 committed:
format: reverted formatting change

4lch4 committed:
format: removed extra commas

4lch4 committed:
chore: removed unneeded comment

4lch4 committed:
feat: added `getAgeHrs` function

This commit adds the `getAgeHrs` function which will return the age/duration between two `time.Time` objects in hours.

4lch4 committed:
chore: updated go.mod/sum

4lch4 committed:
revert: reverted a formatting change

4lch4 committed:
test: added tests for new functionality

4lch4 committed:
Merge branch 'main' into more-github-metrics

4lch4 commented:
I resolved the merge conflicts so it should be ready to review :)

4lch4 committed:
Merge branch 'main' into more-github-metrics

4lch4 committed:
refactor: renamed TestGetDurationHrs to TestGetAgeHrs

This commit renames the test function to be in line with the function it's testing.

adrielp commented on receiver/gitproviderreceiver/documentation.md:
```diff
@@ -34,6 +34,94 @@ Number of repositories in an organization
 
+### git.repository.pull_request.approval.time
+
+The amount of time for a PR to be approved
```
```suggestion
The amount of time it took for a pull request to be approved.
```

adrielp commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/helpers.go:
```diff
@@ -173,3 +175,40 @@ func (ghs *githubScraper) getContributorCount(
+	return pullRequests, nil
+}
+
+// Get the age/duration between two times in hours. The hours will be rounded up
```
I'm thinking that it'd be better to convert to hours at the visualization/query level and the time/age value emitted from the receiver should be in seconds. The semantic conventions seem to leave it open in the case of time for metrics. 

> time - an instrument that measures passage of time should be called entity.time. For example, system.cpu.time with attribute state = idle | user | system | .... time measurements are not necessarily wall time and can be less than or greater than the real wall time between measurements.
> time instruments are a special case of usage metrics, where the limit can usually be calculated as the sum of time over all attribute values. utilization for time instruments can be derived automatically using metric event timestamps. For example, system.cpu.utilization is defined as the difference in system.cpu.time measurements divided by the elapsed time and number of CPUs.

This is primarily focused at time in the context of system metrics. I think just recording the real time in s/ms would be fine. Wouldn't mind knowing your thoughts on this @atoulme / @astencel-sumo

kentquirk replied:
http semconv just forced request duration to seconds from ms, so it seems like using seconds is the right answer.

4lch4 replied:
The latest commits address this and we now emit seconds instead of hours.

adrielp commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -13,40 +13,85 @@ status:
   repository.name:
-      description: The full name of the Git repository
-      type: string
+    description: The full name of the Git repository
```
```suggestion
    description: The name of the Git repository
```

4lch4 replied:
I believe it's just a matter of whether or not the owner of the repo is included. For example, in the case of the current repo we're in, the name would be `opentelemetry-collector-contrib` whereas the full name is `open-telemetry/opentelemetry-collector-contrib`.

adrielp commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -13,40 +13,85 @@ status:
+    description: The full name of the Git repository
+    type: string
+  branch.name:
+    description: The name of the branch in a given repository
```
```suggestion
    description: The name of a Git branch
```

adrielp commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -13,40 +13,85 @@ status:
+    gauge:
+      value_type: int
+    attributes: [repository.name, branch.name]
+  git.repository.pull_request.merge.time:
```
When I read this, I think: "why would we want to know the age of a pull request that has been merged?"

4lch4 replied:
Unless I'm misunderstanding, I believe it's so you can see roughly how long your average PR takes to get merged in a historical view.

4lch4 committed:
Update receiver/gitproviderreceiver/metadata.yaml

Co-authored-by: Adriel Perkins <adrielp@liatrio.com>

4lch4 committed:
[chore] resolve conflicts for merge

4lch4 committed:
[chore] fixed spelling mistake

4lch4 committed:
[chore] added changelog file

4lch4 committed:
[chore] ran generate/docs commands

4lch4 committed:
[refactor] getAgeHrs is now just getAge

The `getAgeHrs` function has been refactored to just be `getAge` and will return the age in seconds as opposed to hours.

4lch4 committed:
[tests] added missing time.Duration for tests

4lch4 committed:
[refactor] removed git.repository.pull_request.deployment.time metric

4lch4 committed:
[refactor] updated metric descriptions and renamed for clarity

4lch4 committed:
[fix] updated function calls for new metric names

adrielp commented on .chloggen/more-github-metrics.yaml:
```diff
@@ -0,0 +1,35 @@
+# These lines will be padded with 2 spaces and then inserted directly into the document.
+# Use pipe (|) for multiline entries.
+subtext: |
+  This is the 3rd PR for the gitproviderreceiver that adds functionality for the following Pull Request metrics:
```
We don't need to callout that it's the third pr. We can just leave the list of metrics.

adrielp replied:
```suggestion
```

adrielp commented on .chloggen/more-github-metrics.yaml:
```diff
@@ -0,0 +1,35 @@
+  - git.repository.pull_request.merged.count
+  - git.repository.pull_request.merge.time
+  - git.repository.pull_request.approval.time
+  - git.repository.pull_request.deployment.time
```
This metric has been removed per our pairing session and should be removed here.

adrielp replied:
```suggestion
```

adrielp commented on receiver/gitproviderreceiver/documentation.md:
```diff
@@ -34,6 +34,79 @@ Number of repositories in an organization
+
+| Name | Description | Values |
+| ---- | ----------- | ------ |
+| repository.name | The full name of a Git repository | Any Str |
```
I think we agreed to not say `the full name` and instead just say `The name of a Git repository`.

adrielp commented on receiver/gitproviderreceiver/documentation.md:
```diff
@@ -34,6 +34,79 @@ Number of repositories in an organization
+
+| Name | Description | Values |
+| ---- | ----------- | ------ |
+| repository.name | The full name of a Git repository | Any Str |
```
I think we agreed to not say `the full name` and instead just say `The name of a Git repository`.

adrielp commented on receiver/gitproviderreceiver/documentation.md:
```diff
@@ -34,6 +34,79 @@ Number of repositories in an organization
+
+| Name | Description | Values |
+| ---- | ----------- | ------ |
+| repository.name | The full name of a Git repository | Any Str |
```
```suggestion
| repository.name | The name of a Git repository | Any Str |
```

4lch4 committed:
[refactor] reworded description of repository.name

adrielp commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -13,40 +13,78 @@ status:
   repository.name:
-      description: The full name of the Git repository
-      type: string
+    description: The full name of a Git repository
```
```suggestion
    description: The name of a Git repository
```

4lch4 replied:
This was just resolved, not sure how this one slipped through the cracks.

adrielp commented on .chloggen/more-github-metrics.yaml:
```diff
@@ -0,0 +1,35 @@
+component: gitproviderreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Added Pull Request metrics to the gitproviderreceiver.
```
```suggestion
note: Add pull request metrics
```

adrielp commented on .chloggen/more-github-metrics.yaml:
```diff
@@ -0,0 +1,35 @@
+# Use pipe (|) for multiline entries.
+subtext: |
+  This is the 3rd PR for the gitproviderreceiver that adds functionality for the following Pull Request metrics:
+
```
```suggestion
```

4lch4 committed:
Apply suggestions from @adrielp in code review

Co-authored-by: Adriel Perkins <adrielp@liatrio.com>

astencel-sumo commented on .chloggen/more-github-metrics.yaml:
```diff
@@ -0,0 +1,32 @@
+  - git.repository.pull_request.open.count
+  - git.repository.pull_request.open.time
+  - git.repository.pull_request.merged.count
+  - git.repository.pull_request.merge.time
```
```suggestion
  - git.repository.pull_request.merged.time
```

adrielp commented on .chloggen/more-github-metrics.yaml:
```diff
@@ -0,0 +1,32 @@
+  - git.repository.pull_request.open.time
+  - git.repository.pull_request.merged.count
+  - git.repository.pull_request.merge.time
+  - git.repository.pull_request.approval.time
```
```suggestion
  - git.repository.pull_request.approved.time
```

4lch4 committed:
Apply suggestions from code review

Co-authored-by: Andrzej Stencel <astencel@sumologic.com>
Co-authored-by: Adriel Perkins <adrielp@liatrio.com>

astencel-sumo commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:
```diff
@@ -122,13 +120,50 @@ func (ghs *githubScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+					ghs.mb.RecordGitRepositoryPullRequestMergedTimeDataPoint(now, age, name, pr.HeadRefName)
+
+				} else {
+					open++
```
What about PRs that were not merged but were instead closed without merging? Are they counted as open here?

astencel-sumo replied:
Oh I think I see, in `helpers.go` we're only retrieving Open and Merged PRs, so we disregard Closed PRs.

4lch4 replied:
I see this is marked as resolved, but just in case, that's correct. We only deal with Open and Merged PRs.

adrielp reviewed (approved):
Approved. Just need to get the workflows to run, pull in latest from main, and get 1 more approval.

atoulme commented:
```
Vulnerability #1: GO-2023-2402
    Man-in-the-middle attacker can compromise integrity of secure channel in
    golang.org/x/crypto
  More info: https://pkg.go.dev/vuln/GO-2023-2402
  Module: golang.org/x/crypto
    Found in: golang.org/x/crypto@v0.16.0
    Fixed in: golang.org/x/crypto@v0.17.0
    Example traces found:
Error:       #1: libpod_client.go:22:30: podmanreceiver.init calls fmt.Errorf, which eventually calls ssh.channel.Close
Error:       #2: podman_connection.go:25:2: podmanreceiver.init calls ssh.init, which eventually calls ssh.channel.Read
Error:       #3: libpod_client.go:22:30: podmanreceiver.init calls fmt.Errorf, which eventually calls ssh.channel.Write
```
This is fixed in main and you can pick up the fix by rebasing. This shouldn't stop a merge as this is not part of the required build.

4lch4 committed:
Merge branch 'main' into more-github-metrics

4lch4 commented:
I just went ahead and ran update to merge changes from `main`.

astencel-sumo commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -13,40 +13,78 @@ status:
+    gauge:
+      value_type: int
+    attributes: [repository.name]
+  git.repository.pull_request.open.time:
```
I'm not convinced that `open` is a good namespace to create in the `git.repository.pull_request` namespace, especially if we drop `git.repository.pull_request.open.count` in favor of `git.repository.pull_request.count`.

How about `git.repository.pull_request.open_time`?

Actually, this component is in `development` status, so we don't need to sweat the naming too hard, we can always introduce breaking changes. 😉

4lch4 replied:
Yeah, I think it'd be best to change this in another PR so we don't clog this one up too much, it's already a pretty big one.

astencel-sumo commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -13,40 +13,78 @@ status:
+    gauge:
+      value_type: int
+    attributes: [repository.name, branch.name]
+  git.repository.pull_request.approved.time:
```
I would get rid of the `approved` namespace. How about `git.repository.pull_request.time_to_approval`?

4lch4 replied:
Same as my comment above, I agree with the change but wondering if we should do this in the next PR?

astencel-sumo commented on receiver/gitproviderreceiver/internal/scraper/githubscraper/github_scraper.go:
```diff
@@ -122,13 +120,50 @@ func (ghs *githubScraper) scrape(ctx context.Context) (pmetric.Metrics, error) {
+
+					age := getAge(pr.CreatedAt, now.AsTime())
+
+					ghs.mb.RecordGitRepositoryPullRequestOpenTimeDataPoint(now, age, name, pr.HeadRefName)
```
So depending on the pull request's status (merged or not), some metrics are emitted and some are not:
- `git.repository.pull_request.merged.time` is only emitted if the pull request is merged - I guess this makes sense 😄 
- `git.repository.pull_request.open.time` is only emitted if the pull request is open - I'm not sure about this
- `git.repository.pull_request.approved.time` is only emitted if the pull request is approved and still open - I'm not sure about this either.

Anyway, this component is in `development` stability, so I guess we should just merge, evaluate how this can be used, and adjust accordingly.

Can you describe the way(s) you intend to use (or are using) these metrics? That would help a lot reasoning what makes sense.

4lch4 replied:
> - `git.repository.pull_request.open.time` is only emitted if the pull request is open - I'm not sure about this
> - `git.repository.pull_request.approved.time` is only emitted if the pull request is approved and still open - I'm not sure about this either.

@adrielp and I talked about these when we redid the metrics last week and made this a conscious choice, but now I'm leaning more toward your take that it might be useful to emit no matter what.

astencel-sumo replied:
Understanding how these metrics are/will be used would help a lot here.

adrielp replied:
Sorry for the delay in response to this, I've been away for the holidays and completely mentally checked out :D 


I like the comment about using state as an attribute and having these all be under the `pull_request` namespace. 

Time here really means `age`, but time is preferred name in the specification which is why we chose it.

Merged, open, and approved are all state, and thus when they change, they're essentially a different metric.

These metrics are used to help teams understand where their bottle necks are and how they could improve if they so choose.  With pull request metrics for example, a team would be able to understand what their lead time is for getting pull requests and contributions through the their code base.

If they have a lot of pull requests in the open state, and the time of which they've been open is long, then it's very likely the team is not pairing, or they have to few people available for PR reviews. 

And that might be ok for a team. But it also might be an indicator that processes should change. They're also not metrics to be used in isolation. They're meant to be used with the other metrics to form a picture.  

We have a whole doc on leading indicators and how teams could leverage these metrics that we've been working on publishing publicly. My intent (which hasn't made it in here yet) is to link & reference that guidance on these metrics within this receiver. Additionally, I'd include a link to the demo that runs and displays the visualization of these metrics.

I'd vote for merging as-is (we know these metrics work at this time in point) and make quick small changes to the names in subsequent PR's while in development.

astencel-sumo replied:
Awesome, I love your answer @adrielp , thank you for this.

> Sorry for the delay in response to this, I've been away for the holidays and completely mentally checked out 

That's completely OK and the right thing to do!

> We have a whole doc on leading indicators and how teams could leverage these metrics that we've been working on publishing

This sounds great, I definitely want to read it when it's public.

> I'd vote for merging as-is (we know these metrics work at this time in point) and make quick small changes

Makes sense to me, again - given that the component is in Development stability status, I suppose that's the way to go.

One last thing I'm waiting for (after the holiday season ends for everybody 🙂) is to update the README, I'm thinking about the GitHub Metrics section.﻿
![image](https://github.com/open-telemetry/opentelemetry-collector-contrib/assets/70892616/5e5a8f9a-5bb0-4a3f-b17a-7701223dca8b)

adrielp replied:
Sounds good to me! Agreed on updating that readme. Hopefully that can be updated tomorrow & we can get this merged this so we can get some small pr's open to iterate! Appreciate the thorough review on this and insight!

4lch4 replied:
There we go, I just updated the README to fix this.

andrzej-stencel reviewed (approved):
Probably worth adjusting the metric units to conform with the [spec](https://github.com/open-telemetry/semantic-conventions/blob/v1.24.0/docs/general/metrics.md#instrument-units).

I have some comments regarding metric names and the actual usage of the metrics, but since the component is in `development` stability, I suppose the best way to iron out the details is to merge and improve incrementally.

astencel-sumo commented:
Wait, I think we also need an update to the README, at least to signify what metrics are available.

4lch4 commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -13,40 +13,78 @@ status:
+    gauge:
+      value_type: int
+    attributes: [repository.name]
+  git.repository.pull_request.merged.count:
```
Oooo.. I quite like that and hadn't thought of it. I'll put this on my TODO list and get it done ASAP.

4lch4 replied:
@adrielp, thoughts?

adrielp replied:
I like this idea. I would prefer to see it done as a change while this is in development.

4lch4 committed:
Apply suggestions @astencel-sumo made in code review

Co-authored-by: Andrzej Stencel <astencel@sumologic.com>

4lch4 commented on receiver/gitproviderreceiver/metadata.yaml:
```diff
@@ -13,40 +13,78 @@ status:
+    gauge:
+      value_type: int
+    attributes: [repository.name, branch.name]
+  git.repository.pull_request.merged.time:
```
I quite like this as well, mind if we do this in the next PR or should I go ahead and fix this up?

4lch4 committed:
chore: updated deps and generated docs w/ new changes

4lch4 commented:
@astencel-sumo, I just ran the usual "generation" commands and `go tidy` to make sure everything is up-to-date with the changes you suggested and I applied.

4lch4 committed:
chore: updated go.mod & go.sum files

4lch4 committed:
chore: updated metrics in README

4lch4 committed:
chore: updated metrics in README

adrielp commented on receiver/gitproviderreceiver/README.md:
```diff
@@ -28,9 +28,11 @@ The current metrics available via scraping from GitHub are:
+- [x] Repository pull request open count
+- [x] Repository pull request time
+- [x] Repository pull request merge time
+- [x] Repository pull request approval time
```
```suggestion
- [x] Repository pull request approved time
```

adrielp commented on receiver/gitproviderreceiver/README.md:
```diff
@@ -28,9 +28,11 @@ The current metrics available via scraping from GitHub are:
+- [x] Repository pull request merge count
+- [x] Repository pull request open count
+- [x] Repository pull request time
+- [x] Repository pull request merge time
```
```suggestion
- [x] Repository pull request merged time
```

4lch4 committed:
chore: updated metric description in README

4lch4 committed:
chore: updated metrics description in README

4lch4 committed:
refactor: replaced unit names with numbers

4lch4 committed:
chore: updated deps and generated docs w/ new changes

4lch4 commented:
Okay, this PR is finally ready to merge. My last commits reverted the earlier suggested change where units were things like `unit: "{pull_request}" as it broke the receiver. We can add this back later, but I've reverted it for now to get this PR done.

astencel-sumo commented on receiver/gitproviderreceiver/documentation.md:
```diff
@@ -34,6 +34,79 @@ Number of repositories in an organization
+
+| Unit | Metric Type | Value Type |
+| ---- | ----------- | ---------- |
+| 1 | Gauge | Int |
```
Looks like all the metric unit fixes have been lost somewhere in the rebases/merges? See [this comment](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/29642#pullrequestreview-1794151974) on this PR and the changes below that comment.

```suggestion
| s | Gauge | Int |
```

4lch4 committed:
chore: reverting a reverted commit as it wasnt an issue

4lch4 committed:
chore: updated deps and generated docs w/ latest changes

andrzej-stencel reviewed (approved):
:rocket: :clap: :bow:

4lch4 committed:
Merge branch 'main' into more-github-metrics

4lch4 committed:
Merge remote-tracking branch 'upstream/main' into more-github-metrics

4lch4 committed:
[chore] fixed conflict and ran updates

4lch4 committed:
Merge branch 'main' into more-github-metrics

adrielp commented:
Looks like the build-and-test for windows is failing because it's trying to test a podman receiver on windows. This shouldn't block the merge of this pr correct @atoulme @astencel-sumo ?

4lch4 committed:
Merge branch 'main' into more-github-metrics

Commit:
Merge branch 'main' into more-github-metrics

Commit:
chore: run tidy

Commit:
Merge branch 'main' into more-github-metrics

adrielp commented:
Have resolved several conflicts after changes to main and changes for lifecycle tests being merged. Workflows have been re-run now and the unrelated windows workflows that were previously failing look to be fixed. We're good to go!
